lib/srv/db/redis/engine.go

```diff
@@ -261,26 +246,101 @@ func (e *Engine) getNewClientFn(ctx context.Context, sessionCtx *common.Session)
 
 // createOnClientConnectFunc creates a callback function that is called after a
 // successful client connection with the Redis server.
-func (e *Engine) createOnClientConnectFunc(sessionCtx *common.Session, username, password string) onClientConnectFunc {
+func (e *Engine) createOnClientConnectFunc(ctx context.Context, sessionCtx *common.Session, username, password string) onClientConnectFunc {
 	switch {
 	// If password is provided by client.
 	case password != "":
 		return authWithPasswordOnConnect(username, password)
 
-	// If database user is one of managed users.
+	// Azure databases authenticate via access keys.
+	case sessionCtx.Database.IsAzure():
+		credFetchFn := azureAccessKeyFetchFunc(sessionCtx, e.Auth)
+		return fetchCredentialsOnConnect(e.Context, sessionCtx, e.Audit, credFetchFn)
+
+	// If database user is one of managed users (AWS only).
 	//
 	// Teleport managed users can have their passwords rotated during a
 	// database session. Fetching an user's password on each new connection
 	// ensures the correct password is used for each shard connection when
 	// Redis is in cluster mode.
 	case slices.Contains(sessionCtx.Database.GetManagedUsers(), sessionCtx.DatabaseUser):
-		return fetchUserPasswordOnConnect(sessionCtx, e.Users, e.Audit)
+		credFetchFn := managedUserCredFetchFunc(sessionCtx, e.Auth, e.Users)
+		return fetchCredentialsOnConnect(e.Context, sessionCtx, e.Audit, credFetchFn)
+
+	// AWS ElastiCache has limited support for IAM authentication.
+	// See: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/auth-iam.html#auth-iam-limits
+	// So we must check that the database supports IAM auth and that the
+	// ElastiCache user has IAM auth enabled.
+	case e.isAWSIAMAuthSupported(ctx, sessionCtx):
+		credFetchFn := elasticacheIAMTokenFetchFunc(sessionCtx, e.Auth)
+		return fetchCredentialsOnConnect(e.Context, sessionCtx, e.Audit, credFetchFn)
 
 	default:
 		return nil
 	}
 }
 
+// isAWSIAMAuthSupported returns whether AWS IAM auth is supported for the
+// database and the database user.
+func (e *Engine) isAWSIAMAuthSupported(ctx context.Context, sessionCtx *common.Session) bool {
```

Comment:
nit: i believe `onConnect` is called on each new instance connection when Redis is in cluster mode. Ideally, we only do `checkUserIAMAuthIsEnabled` once.

Reply:
you're correct, and the other checks in this function do not incur an API call. I'll try to address this. maybe just memoize the function result?

Reply:
Saved the result of this function: https://github.com/gravitational/teleport/pull/25967/commits/8e0797232374a8e461b6d94dad1099168356e56b